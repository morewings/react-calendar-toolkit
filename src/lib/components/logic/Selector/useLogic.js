import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  subMonth,
  floorMonth,
  addMonth,
} from 'lib/utils/dateUtils';
import {useDatepickerActions} from 'lib/features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatepickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = addMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = subMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
