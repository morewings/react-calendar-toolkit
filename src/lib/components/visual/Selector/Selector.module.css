.wrapper {
  align-items: center;
  display: flex;
  height: var(--selectorHeight);
  justify-content: space-between;
  padding: calc(var(--innerPadding) / 2) var(--innerPadding);
}

.buttons button {
  background-color: transparent;
  border: none;
  border-radius: 4px;
  color: var(--baseColor);
  cursor: pointer;
  font-size: var(--fontSize);
  margin-right: 2px;
  padding: 8px;
  user-select: none;
}

.stepper button {
  background-color: transparent;
  border: none;
  border-radius: 50%;
  color: var(--baseColor);
  cursor: pointer;
  font-size: var(--fontSize);
  font-weight: bold;
  height: 36px;
  padding: 0;
  text-align: center;
  user-select: none;
  width: 36px;
}

.stepper button:active,
.stepper button:focus {
  outline: none;
}

.stepper button:focus,
.stepper button:hover {
  background-color: var(--entryHoverColor);
}

.buttons button:active,
.buttons button:focus {
  outline: none;
}

.buttons button:focus,
.buttons button:hover {
  background-color: var(--entryHoverColor);
}
