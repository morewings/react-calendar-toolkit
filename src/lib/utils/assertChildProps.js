import React, {Fragment} from 'react';

export const trackProps = jest.fn();

export const DummyComponent = props => {
  trackProps(props);
  return <Fragment>Child</Fragment>;
};
