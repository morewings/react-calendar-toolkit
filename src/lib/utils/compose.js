const compose = (...fns) => (...args) =>
  fns.reduceRight(
    (params, f) => (Array.isArray(params) ? f(...params) : f(params)),
    args
  );

export default compose;
