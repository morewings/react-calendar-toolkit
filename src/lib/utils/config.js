export default Object.freeze({
  supportedPrecisions: Object.freeze([
    'year',
    'month',
    'day',
    'hour',
    'minute',
  ]),
});
