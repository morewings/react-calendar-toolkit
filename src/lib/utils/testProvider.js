import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'lib/features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'lib/features/theme/withTheme';
import {Provider as ModalProvider} from 'lib/features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'lib/features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
