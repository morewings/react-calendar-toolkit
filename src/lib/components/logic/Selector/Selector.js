import React from 'react';
import PropTypes from 'prop-types';
import {convertToDate} from 'lib/utils/dateUtils';
import {useMonthStepperLabels} from 'lib/features/locale';
import useLogic from './useLogic';

const Selector = ({
  todayTimestamp,
  selectedTimestamp,
  renderAs,
  startDate,
  endDate,
  visibleTimestamp,
  precision,
}) => {
  const monthStepperLabels = useMonthStepperLabels();

  const {onSetPrecision, onIncrementMonth, onDecrementMonth} = useLogic({
    startDate,
    endDate,
    visibleTimestamp,
  });

  const SelectorVisual = renderAs;
  return (
    <SelectorVisual
      precision={precision}
      monthStepperLabels={monthStepperLabels}
      incrementMonth={onIncrementMonth}
      decrementMonth={onDecrementMonth}
      setPrecision={onSetPrecision}
      todayDate={convertToDate(todayTimestamp)}
      selectedDate={convertToDate(selectedTimestamp)}
      visibleDate={convertToDate(visibleTimestamp)}
    />
  );
};

Selector.propTypes = {
  selectedTimestamp: PropTypes.number.isRequired,
  todayTimestamp: PropTypes.number.isRequired,
  visibleTimestamp: PropTypes.number.isRequired,
  renderAs: PropTypes.elementType.isRequired,
  startDate: PropTypes.instanceOf(Date).isRequired,
  endDate: PropTypes.instanceOf(Date).isRequired,
  precision: PropTypes.oneOf(['year', 'month', 'day']).isRequired,
};

export default Selector;
