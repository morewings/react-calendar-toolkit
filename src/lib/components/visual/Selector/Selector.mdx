---
name: Selector
menu: Customize UI
---
import { Playground, Props } from 'docz';
import Datepicker from 'lib';
import Selector from './Selector';

`Selector` component is rendered to the top of calendar. It allows changing [calendar precision](/#calendar-views-and-precision) (`setPrecision`) and stepping months (`incrementMonth`, `decrementMonth`). Replace it with custom component.

<Playground>
  {() => {
    const style = {
      color: 'black',
      textAlign: 'center',
      padding: '16px 0',
      border: '2px dashed pink',
      margin: '2px'
    };
    const CustomComponent = () => {
      return (
        <div style={style}>
          {`<Selector />`}
        </div>
      )
    };
    return (
      <Datepicker
        onDateSet={date => {console.log('date set', date);}}
        minPrecision="month"
        renderSelectorAs={CustomComponent}
        title="Demo datepicker" />
    )
  }}
</Playground>

## Props

<Props of={Selector} />

