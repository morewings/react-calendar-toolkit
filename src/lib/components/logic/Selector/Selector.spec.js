import React from 'react';
import {render} from '@testing-library/react';
import {DatepickerMockProvider as Provider} from 'lib/utils/testProvider';
import Selector from './Selector';

const MockComponent = jest.fn(() => <div>MockComponent</div>);

const renderWithProviderProps = (Component, props = {}) =>
  render(<Component />, {
    wrapper: ({children}) => <Provider {...props}>{children}</Provider>,
  });

const date = new Date(2012, 11, 12); // 12.12.2012

describe('Selector', () => {
  beforeEach(() => {
    MockComponent.mockClear();
  });

  it('renders', () => {
    const props = {
      renderAs: MockComponent,
      todayTimestamp: date.getTime(),
      selectedTimestamp: date.getTime(),
      startDate: new Date(2011, 11, 12),
      endDate: new Date(2015, 11, 12),
      visibleTimestamp: date.getTime(),
      precision: 'day',
    };

    const {asFragment} = renderWithProviderProps(() => <Selector {...props} />);
    expect(asFragment()).toMatchSnapshot();
    expect(MockComponent.mock.calls[0][0]).toMatchSnapshot();
  });
});
