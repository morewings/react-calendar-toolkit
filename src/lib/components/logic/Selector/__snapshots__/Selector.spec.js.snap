// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector renders 1`] = `
<DocumentFragment>
  Child
</DocumentFragment>
`;

exports[`Selector renders 2`] = `
Object {
  "decrementMonth": [Function],
  "incrementMonth": [Function],
  "monthStepperLabels": Object {
    "decrementMonthLabel": "1 month ago",
    "incrementMonthLabel": "in 1 month",
  },
  "precision": "day",
  "selectedDate": 2012-12-12T00:00:00.000Z,
  "setPrecision": [Function],
  "todayDate": 2012-12-12T00:00:00.000Z,
  "visibleDate": 2012-12-12T00:00:00.000Z,
}
`;
